crates/optimism/rpc/src/historical.rs

```diff
@@ -145,8 +153,95 @@ where
         })
     }
 
-    fn batch<'a>(&self, req: Batch<'a>) -> impl Future<Output = Self::BatchResponse> + Send + 'a {
-        self.inner.batch(req)
+    fn batch<'a>(
+        &self,
+        mut req: Batch<'a>,
+    ) -> impl Future<Output = Self::BatchResponse> + Send + 'a {
+        use futures::future::{BoxFuture, FutureExt};
+
+        let inner_service = self.inner.clone();
+        let historical = self.historical.clone();
+
+        async move {
+            let mut needs_forwarding = false;
+            for entry in req.iter_mut() {
+                if let Ok(BatchEntry::Call(call)) = entry &&
+                    historical.should_forward_request(call)
+                {
+                    needs_forwarding = true;
+                    break;
+                }
+            }
+
+            if !needs_forwarding {
+                return inner_service.batch(req).await;
+            }
+
+            let entries: Vec<_> = req.into_iter().collect();
```

Comment:
maybe we can simplify this a bit and simply flatten all requests here and use self.call ?